mllib/src/main/scala/org/apache/spark/ml/Predictor.scala

```diff
@@ -62,6 +62,40 @@ private[ml] trait PredictorParams extends Params
     }
     SchemaUtils.appendColumn(schema, $(predictionCol), DoubleType)
   }
+
+  /**
+   * Extract [[labelCol]], weightCol(if any) and [[featuresCol]] from the given dataset,
+   * and put it in an RDD with strong types.
+   */
+  protected def extractInstances(dataset: Dataset[_]): RDD[Instance] = {
+    val w = this match {
+      case p: HasWeightCol =>
+        if (isDefined(p.weightCol) && $(p.weightCol).nonEmpty) {
+          col($(p.weightCol)).cast(DoubleType)
+        } else {
+          lit(1.0)
+        }
+      case _ => lit(1.0)
```

Comment:
If it doesn't have a weight column, does it mean there's no point in selecting lit(1.0) as a weight column as it will be unused? or do some algorithms not have a weight column but nevertheless have ways of using a weight?

Reply:
You are right, if an alg do not have `weightCol`, it should not deal with weighting.
So, what about raising an exception instead of assign it to 1.0?

Reply:
I think an error will occur elsewhere? is it necessary to handle it here vs just not making an empty col?

Reply:
Since this method will only be called internally, so I think it is update to the developers to decide whether to use it or not. If an algorithm (like GBT) do not support weighting now, it can use existing `extractLabeledPoints` instead.

Reply:
OK, I suppose I'm just concerned about the small overhead of adding an unused column.
You're saying that it's up to implementations to call the method they need, one with weights or not? yeah I agree, and they will call the right method in this change? If true, then do you even need this check? it will already fail (correctly) if the code is calling the wrong method.

Reply:
OK, I will remove this line. Impls that do not support weighting call this method should fail.